### DISCUSSION TITLE
[Build PupilEXT from Source]: Linking with Pylon 6 library on Windows 8.1 (x64) failed in VSCode

### DISCUSSION THREAD

**Discussion by kheki4** [openPupil/Open-PupilEXT#21] [Q&A] [answered]
Hi everyone!

I am now trying to build the program from the provided source code. The machine is running Windows 8.1 x64. I am trying to build with MSVC, "Visual Studio Build Tools 2017 Release - amd64" kit. Build halts when trying to link the pylon library:
"...
[build] LINK : fatal error LNK1104: cannot open file 'GCBase_MD_VC141_v3_1_Basler_pylon.lib' [C:\Open-PupilEXT\build\src\PupilEXT.vcxproj]
[build] Build finished with exit code 1 
"
First when I encountered this error, I believed the problem might have been that the path had whitespace characters, like "C:\Program Files\..." so I reinstalled the pylon software into "C:\basler-pylon-6".. and changed the corresponding lines in the CMakeLists.txt of the pupilEXT root directory to this:
" ...
elseif(MSVC OR WIN32)
    set(PYLON_HOME "C:/basler-pylon-6/Development")
    set(PYLON_INCLUDE_DIR "C:/basler-pylon-6/Development/include")
... "
But the error message is still the same, and the program cannot compile.
Can you please help? 
I am new to using VS Code and these methods of configuring the build procedure. So far I have only worked with simple Visual C# projects for university homeworks, where I did not have to manage libraries like this. 
I am a masters student in cognitive science and only have considerable experience with handling preprocessed pupil data in Matlab. Now I am trying to understand and learn how pupil detection and data filtering works on this lower level, that is why I am compiling PupilEXT myself (also hoping to use the program with the camera if I can arrange the purchase for a setup, later this year).

Thanks in advance!

**Comment by BZandi** [comment]
There is no need to change the path of Pylon. It is recommended to install Pylon in the program files folder as in the provided [cmake](https://github.com/openPupil/Open-PupilEXT/blob/main/CMakeLists.txt) file the standard path is already adjusted for you. Furthermore, as vcpkg manifest is used for the libs, everything should work if you follow the tutorial. Please follow every step described in the tutorial (see [here](https://github.com/openPupil/Open-PupilEXT)), which is provided in the section "3.0 Build PupilEXT with vcpkg manifest (recommended)".

If only the link failed (at the very end of the build), it might have to do with the fact that the Pylon libs are not adjusted into your system variables. See the attached image how it should look like.
![Env-variables](https://user-images.githubusercontent.com/62992112/153649472-98f2f92c-e093-426d-961f-41dc245327b3.png)

On this page ([Link](https://docs.oracle.com/en/database/oracle/machine-learning/oml4r/1.5.1/oread/creating-and-modifying-environment-variables-on-windows.html#GUID-DD6F9982-60D5-48F6-8270-A27EC53807D0)), there is a tutorial on how adding a folder to the system variables of windows. In the section "system variables", go to the line "path", press edit and add the Pylon folder as shown in the upper image to the other listed elements. It should work.

**Reply by kheki4** [reply]
Okay. I reinstalled Pylon 6 againt, just to make sure that is is located in the default folder.
As I saw later, the installer successfully created the variable PYLON_DEV_DIR among system variables with value:
C:\Program Files\Basler\pylon 6\Development\
and added these to the Path system variable:
C:\Program Files\Basler\pylon 6\Runtime\x64\;
C:\Program Files\Basler\pylon 6\Runtime\Win32\

Then I modified CMakeLists.txt to the original:
"...
elseif(MSVC OR WIN32)
	set(PYLON_HOME "C:/Program Files/Basler/pylon 6/Development")
    set(PYLON_INCLUDE_DIR "C:/Program Files/Basler/pylon 6/Development/include")
..."

Then I restarted VS Code.
And I ran the build command again, but the error message remains the same.
I also rebooted the computer and restarted the build again, but still the same error occurs.

**Reply by BZandi** [reply]
Please provide all messages that appear in your terminal after you press compile in your editor so that I can check the build process. Sometimes it is necessary to delete the previous build folder to start a clean build, as the remaining CMake cache files might produce errors. 

**Edit:**
Please also verify whether the `GCBase_MD_VC141_v3_1_Basler_pylon.lib` exists in your `C:\Program Files\Basler\pylon 6\Development\lib\x64` folder.

**Reply by kheki4** [reply]
Now I deleted the entire source code folder, and downloaded it fresh from github, started to build again, but the problem appears still. (I also tried when running VS Code "As administrator", but is did not make a change.) Here is the build output log, and a cmd printscreen of the Basler library folder:
[build-log.txt](https://github.com/openPupil/Open-PupilEXT/files/8052943/build-log.txt)
![lib-location](https://user-images.githubusercontent.com/33413678/153700272-327d9dce-64ad-44d5-80fc-ccd287e0ac75.png)
.

**Reply by BZandi** [reply]
It needs to be clarified whether the issue is that the lib cannot be found or there is another source of trouble caused by your compiler. The advantage of windows is that libs will be searched automatically in the root folder of the *.exe file before looking into other directories. Try the following steps:

1) I saw that you built the project in **debug mode**. Please remove the build folder in the PupilEXT directory and build the project again but use the **release mode**. After you build the project in **release mode**, the error will occur again, but there should be the new folder `build/src/Release` in which you should find the file `PupilEXT.exe` (see attached image).

2) Now you can copy the lib files located in the `C:\Program Files\Basler\pylon 6\Development\lib\x64`into the `build/src/Release` folder. Then, double click on the PupilEXT.exe file or build the project again. What happens?

![Build folder](https://user-images.githubusercontent.com/62992112/153706368-24d9b62e-7dc0-4790-bb5c-4363f4020e98.png)

**Edit:**
I saw that you have an underscore (`_`) in your path `C:\_pupillometry\Open-PupilEXT`. Please remove it so that it looks like `C:\pupillometry\Open-PupilEXT`. Did you start vscode with admin right ?

**Reply by kheki4** [reply]
I tried out your suggestions, actually there I found no .exe file in the Release folder, and the building process did not go further than the same error again, even with the lib files copied there. Then I tried with copying the lib files into the build/lib folder, which caused the project to compile completely, but the compiled application could not run, because it did not find the Pylon v6.0 dll file. Then I realized that I was wrong all along. **I mistakenly installed Basler Camera software suite v6.3.0 instead of v6.0.1, that was the real reason!** I am sorry, it was only my fault. In the building manual description, you have clearly stated: "We tested PupilEXT with Pylon 6.0.1. on several fresh installed Windows 10 (64 bit) machines. Only the Pylon 6.0.1. version works with PupilEXT, as Windows is a diva." As the major version was still 6, I just somehow did not pay attention for the minor version numbers. Excuse me for your time.

Anyway, before the linking error problem, I have already experienced 3 difficulties in the past. Now I describe them and their solutions, as they may be important for other people, who try compiling this source:
1. CMake-tools could not find CMake, even if it was in my Windows Path system variable. At last, I figured out that I have to modify .vscode/settings.json to this, to solve this problem:
{
    "cmake.configureSettings": {
        "VCPKG_TARGET_TRIPLET": "x64-windows-static-md",
        "CMAKE_TOOLCHAIN_FILE": "3rdparty/vcpkg/scripts/buildsystems/vcpkg.cmake"
    },
    "cmake.cmakePath": "C:/CMake/bin/cmake.exe"
}
2. During building configuration step, as vcpkg downloaded some missing zip file(s), it tried to unzip them, but could not do it. It said, it could not find 7zip on the computer... which is strange, as I have my 7zip installed, even in a non-whitespace path on my C drive. Then it always tried to download its .msi file to use it locally. I reinstalled 7zip, and also added its path to the Path system environment variable. I could call "7z.exe" from cmd anywhere, so it the variable was working, however vcpkg still could not find it. Then I downloaded "7z1900-x64.msi" from the internet manually, and put it right into the folder: 3rdparty/vcpkg/downloads , and then it started to work fine.
3. Then (as I was first trying to compile using mingw32 GCC 6.3.0 compiler) vcpkg said: "No suitable Visual Studio instances were found" which turned out to say "I need MSVC, and only a specific version please" in human words. So after a little googling, I doenloaded "vs_BuildTools.exe" and from that I installed "Visual Studio Build Tools 2017" with additional package "Visual C++ Tools for CMake". And then I needed to manually add a Windows system environment variable with this name: "VCPKG_VISUAL_STUDIO_PATH" and value "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools". Then build could continue further, now using the newly added compiler kit "Visual Studio Build Tools 2017 - amd64".

+ Note: VS Code User Installer version also produced some problem (I actually do not remember the error message at this moment), and it turned out, that System Installer version is the working one, so I had to reinstall VS Code with that. (For example User Installer exe is named like: VSCodeUserSetup-x64-1.64.2.exe, and System Installer is named like: VSCodeSetup-x64-1.64.2.exe)

Now the produced .exe is quite large in size (~11 mb), but it works fine, when I put it among the necessary dll files in the same folder :)

Thank you for all your help! :)

**Comment by kheki4** [comment]
I mistakenly installed Basler Camera software suite v6.3.0 instead of v6.0.1, that was the reason! 
(Further details are in comments under Zandi's post.)

**Reply by BZandi** [reply]
Glad to hear that it works now on your machine! You can also give the editor QT Creator a try instead of VSCode as it is easier to handle and to debug for QT C++ projects.

**Comment by BZandi** [comment, marked as answer]
I checked again the source code of PupilEXT because usually, it should not matter which pylon version you use if you build PupilEXT from the source. The statement in the readme about using Pylon 6.0.1. was just for the pre-compiled binaries.

It turns out that the problem relies on the [``Open-PupilEXT/cmake/FindPylon.cmake``](https://github.com/openPupil/Open-PupilEXT/blob/main/cmake/FindPylon.cmake) file, which is used to find the pylon installation on your local PC when building from the source. In the file's lines 40 to 56, there is the following code

```Cmake
set(PYLON_LIB_SEARCH_PATH "${PYLON_HOME}/lib/x64")

find_library(PYLON_BASE_LIBRARY
            NAMES PylonBase_v6_0.lib
            PATHS ${PYLON_LIB_SEARCH_PATH})
find_library(PYLON_GCBASE_LIBRARY
            NAMES GCBase_MD_VC141_v3_1_Basler_pylon.lib
            PATHS ${PYLON_LIB_SEARCH_PATH})
find_library(PYLON_GENAPI_LIBRARY
            NAMES GenApi_MD_VC141_v3_1_Basler_pylon.lib
            PATHS ${PYLON_LIB_SEARCH_PATH})
find_library(PYLON_UTILITY_LIBRARY
            NAMES PylonUtility_v6_0.lib
            PATHS ${PYLON_LIB_SEARCH_PATH})
find_library(PYLON_GUI_LIBRARY
            NAMES PylonGUI_v6_0.lib
            PATHS ${PYLON_LIB_SEARCH_PATH})
```

As only Pylon 6.0.1 is specified in [``Open-PupilEXT/cmake/FindPylon.cmake``](https://github.com/openPupil/Open-PupilEXT/blob/main/cmake/FindPylon.cmake), the build only works with this version. In other words, you can change the version number in the ’[``Open-PupilEXT/cmake/FindPylon.cmake``](https://github.com/openPupil/Open-PupilEXT/blob/main/cmake/FindPylon.cmake) file manually and then it will be possible to also use Pylon 6.3 installation for the building process.

I will add a remark into the readme about this topic.

**Reply by kheki4** [reply]
Thank you! :)

**Reply by kheki4** [reply]
I finally chose this post as the answer to the initial question, as your description explains the solution to the problem more generally
